pkg/migration/migration.go

```diff
@@ -0,0 +1,438 @@
+		a = append([]interface{}{dc.Namespace, dc.Name}, a...)
+		return imageregistryapi.ImageRegistrySpec{}, nil, fmt.Errorf(format, a...)
+	}
+
+	containers := dc.Spec.Template.Spec.Containers
+	if len(containers) != 1 {
+		return fail("got %d containers, want 1", len("containers"))
+	}
+
+	registry := containers[0]
+	if registry.Name != "registry" {
+		return fail("got container named %q, want %q", registry.Name, "registry")
+	}
+
+	const defaultConfigurationPath = "/config.yml"
+	configurationPath := defaultConfigurationPath
+	hardcodedEnvs := map[string]string{
+		"REGISTRY_HTTP_ADDR": ":5000",
+		"REGISTRY_HTTP_NET":  "tcp",
+	}
+	envs := map[string]string{}
+	for _, env := range registry.Env {
+		if env.ValueFrom != nil {
+			// FIXME(dmage): should we support it?
+			return fail("environment variable %s: valueFrom is not supported", env.Name)
+		}
+		if value, ok := hardcodedEnvs[env.Name]; ok {
+			if env.Value != value {
+				return fail("environment variable %s: got %q, but the only supported value is %q", env.Name, env.Value, value)
+			}
+		} else if env.Name == "REGISTRY_CONFIGURATION_PATH" {
+			configurationPath = env.Value
+		} else if strings.HasPrefix(env.Name, "REGISTRY_") {
+			envs[env.Name] = env.Value
+		} else {
+			return fail("unsupported environment variable %s", env.Name)
+		}
+	}
+
+	// TODO(dmage): registry.EnvFrom?
+
+	podFileGetter := newPodFileGetter(registry, dc.Spec.Template.Spec.Volumes, resources)
+	configData, err := podFileGetter.PodFile(configurationPath)
+	if _, ok := err.(errVolumeMountNotFound); ok {
+		if configurationPath != defaultConfigurationPath {
+			return fail("configuration path %q is not default, but no volume mounts found for it", configurationPath)
+		}
+		configData = []byte(`version: 0.1
+storage:
+  cache:
+    blobdescriptor: inmemory
+  filesystem:
+    rootdirectory: /registry
+  delete:
+    enabled: true
+`)
+	} else if err != nil {
+		return fail("unable to get the configuration file: %s", err)
+	}
+
+	if _, ok := envs["REGISTRY_OPENSHIFT_SERVER_ADDR"]; !ok {
+		envs["REGISTRY_OPENSHIFT_SERVER_ADDR"] = "must-be-set-to-parse-the-config-file"
+	}
+	for key, value := range envs {
+		if err := os.Setenv(key, value); err != nil {
+			return fail("unable to set the environment variable %s: %v", key, err)
+		}
+	}
+	config, extraConfig, err := registryconfig.Parse(bytes.NewReader(configData))
+	if err != nil {
+		return fail("unable to parse the configuration file: %s", err)
+	}
+
+	storage, err := newImageRegistryConfigStorage(config, dc, registry)
+	if err != nil {
+		return fail("unable to make configuration for the storage: %s", err)
+	}
+
+	tls, tlsSecret, err := migrateTLS(config, podFileGetter)
+	if err != nil {
+		return fail("unable to process TLS configuration: %s", err)
+	}
+
+	return imageregistryapi.ImageRegistrySpec{
+		OperatorSpec: operatorapi.OperatorSpec{
+			ManagementState: operatorapi.Managed,
+			Version:         "migrated", // TODO(dmage): what do we need here?
```

Comment:
i'm not sure what we want to use the version field for(or what we'd set it to normally), but it'd say for indicating this ImageRegistry value came from a migration, we should simply annotate it in some way.